firedrake/preconditioners/patch.py

```diff
@@ -1,16 +1,75 @@
-from firedrake.preconditioners.base import PCBase
+from firedrake.preconditioners.base import PCBase, SNESBase
 from firedrake.petsc import PETSc
+from firedrake.solving_utils import _SNESContext
+from firedrake.matrix_free.operators import ImplicitMatrixContext
+from firedrake.dmhooks import get_appctx, push_appctx, pop_appctx
+
 from functools import partial
 import numpy
 import operator
-from ufl import VectorElement, MixedElement
+from ufl import VectorElement, MixedElement, Coefficient, FunctionSpace
+from tsfc.kernel_interface.firedrake import KernelBuilder as FiredrakeKernelBuilder
 
 from pyop2 import op2
 from pyop2 import base as pyop2
 from pyop2 import sequential as seq
 from pyop2.datatypes import IntType
 
-__all__ = ("PatchPC", "PlaneSmoother")
+__all__ = ("PatchPC", "PlaneSmoother", "PatchSNES")
+
+
+class PatchKernelBuilder(FiredrakeKernelBuilder):
+    """Custom kernel interface for patch assembly.
+
+    Ensures that a provided set of coefficients are not split apart if they are mixed.
+
+    This is necessary because PETSc provides the state vector (which
+    may be mixed) as one concatenated vector, rather than the
+    Firedrake convention of one vector per subspace."""
+    def __init__(self, *unsplit_coefficients):
+        self.unsplit_coefficients = frozenset(unsplit_coefficients)
+        self.KernelBuilder = self
+
+    # TSFC makes the kernel builder by calling "interface.KernelBuilder(...)"
+    # on the provided interface object. So mock that with this __call__ method.
+    def __call__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        return self
```

Comment:
I think this `__call__` override is unnecessary (and ugly) since as far as I can see, it is only used so that we can sneak extra data, despite the fact that the class was supposed to be instantiated in TSFC.

A better approach is to have a proper `__init__` function that receives the extra data arguments as well as the ones that will be forwarded to the superclass `__init__`, and then pass the following interface "class" to `compile_form`:
```python
functools.partial(PatchKernelBuilder, unsplit_coefficients=blahblah)
```

Reply:
Okay, so I forgot that TSFC `compile_form` and `compile_interface` in fact expect a Python module as their `interface=` argument, rather than the kernel builder class (or anything that returns a kernel builder instance when called).  However, this module is only used to instantiate its KernelBuilder class, so this is trivial to change.
(Unfortunately, this trivial change would have to be made in FFC as well, if they ever care to adopt updated versions of TSFC.)

A sketch patch to TSFC:
```diff
modified   tsfc/driver.py
@@ -43,8 +43,6 @@ def compile_form(form, prefix="form", parameters=None, interface=None):
 
     assert isinstance(form, Form)
 
-    if interface is None:
-        interface = firedrake_interface
     # Determine whether in complex mode:
     # complex nodes would break the refactoriser.
     complex_mode = parameters and is_complex(parameters.get("scalar_type"))
@@ -70,7 +68,7 @@ def compile_form(form, prefix="form", parameters=None, interface=None):
 
 
 def compile_integral(integral_data, form_data, prefix, parameters,
-                     interface=firedrake_interface):
+                     interface=firedrake_interface.KernelBuilder):
     """Compiles a UFL integral into an assembly kernel.
 
     :arg integral_data: UFL integral data
@@ -109,9 +107,9 @@ def compile_integral(integral_data, form_data, prefix, parameters,
 
     # Dict mapping domains to index in original_form.ufl_domains()
     domain_numbering = form_data.original_form.domain_numbering()
-    builder = interface.KernelBuilder(integral_type, integral_data.subdomain_id,
-                                      domain_numbering[integral_data.domain],
-                                      parameters["scalar_type"])
+    builder = interface(integral_type, integral_data.subdomain_id,
+                        domain_numbering[integral_data.domain],
+                        parameters["scalar_type"])
     argument_multiindices = tuple(builder.create_element(arg.ufl_element()).get_indices()
                                   for arg in arguments)
     return_variables = builder.set_arguments(arguments, argument_multiindices)
modified   tsfc/kernel_interface/firedrake.py
@@ -178,7 +178,8 @@ class ExpressionKernelBuilder(KernelBuilderBase):
 class KernelBuilder(KernelBuilderBase):
     """Helper class for building a :class:`Kernel` object."""
 
-    def __init__(self, integral_type, subdomain_id, domain_number, scalar_type=None):
+    def __init__(self, integral_type, subdomain_id, domain_number,
+                 dont_split=None, scalar_type=None):
         """Initialise a kernel builder."""
         super(KernelBuilder, self).__init__(scalar_type, integral_type.startswith("interior_facet"))
 
@@ -202,6 +203,12 @@ class KernelBuilder(KernelBuilderBase):
         elif integral_type == 'interior_facet_horiz':
             self._entity_number = {'+': 1, '-': 0}
 
+        # Coefficients not to be split
+        if dont_split is None:
+            self.dont_split = ()
+        else:
+            self.dont_split = frozenset(dont_split)
+
     def set_arguments(self, arguments, multiindices):
         """Process arguments.
 
@@ -236,7 +243,7 @@ class KernelBuilder(KernelBuilderBase):
         for i in range(len(integral_data.enabled_coefficients)):
             if integral_data.enabled_coefficients[i]:
                 coefficient = form_data.function_replace_map[form_data.reduced_coefficients[i]]
-                if type(coefficient.ufl_element()) == ufl_MixedElement:
+                if type(coefficient.ufl_element()) == ufl_MixedElement and coefficient not in self.dont_split:
                     split = [Coefficient(FunctionSpace(coefficient.ufl_domain(), element))
                              for element in coefficient.ufl_element().sub_elements()]
                     coefficients.extend(split)

```

Then `dont_split` could be passed on as previously said:

```python
interface = functools.partial(firedrake_interface.KernelBuilder, dont_split=...)
compile_form(..., interface=interface)
```

One might even define a helper function for less clutter:

```python
# this might go in tsfc/kernel_interface/firedrake.py let's say
def make_builder(*args, **kwargs):
    return functools.partial(KernelBuilder, *args, **kwargs)
```

So that the call in firedrake could simply be: `compile_form(..., interface=make_builder(dont_split=...))`